chain/chunks/src/lib.rs

```diff
@@ -178,53 +204,124 @@ impl SealsManager {
-            let chunk_producer = self.runtime_adapter.get_chunk_producer(
-                &self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash)?,
-                height,
-                shard_id,
-            )?;
-            let mut candidates = vec![];
-            for part_ord in 0..self.runtime_adapter.num_total_parts() {
-                let part_ord = part_ord as u64;
-                let part_owner = self.runtime_adapter.get_part_owner(parent_hash, part_ord)?;
-                if part_owner == chunk_producer || Some(part_owner) == self.me {
-                    continue;
-                }
-                candidates.push(part_ord);
+    ) -> Result<Seal, near_chain::Error> {
+        match self.previous_seals.get(&height) {
+            Some(hashes) if hashes.contains(chunk_hash) => Ok(Seal::Previous),
+
+            // None | Some(hashes) if !hashes.contains(chunk_hash)
+            _ => self
+                .get_active_seal(chunk_hash, parent_hash, height, shard_id)
+                .map(|challenge| Seal::Active(challenge)),
+        }
+    }
+
+    fn get_active_seal(
+        &mut self,
+        chunk_hash: &ChunkHash,
+        parent_hash: &CryptoHash,
+        height: BlockHeight,
+        shard_id: ShardId,
+    ) -> Result<&mut ActiveSealChallenge, near_chain::Error> {
+        match self.active_challenges.entry(chunk_hash.clone()) {
+            Entry::Occupied(entry) => Ok(entry.into_mut()),
+            Entry::Vacant(entry) => {
+                let chunk_producer = self.runtime_adapter.get_chunk_producer(
+                    &self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash)?,
+                    height,
+                    shard_id,
+                )?;
+                let candidates = {
+                    let n = self.runtime_adapter.num_total_parts();
+                    // `n` is an upper bound for elements in the accumulator; declaring with
+                    // this capacity up front will mean no further allocations will occur
+                    // from `push` calls in the loop.
+                    let mut accumulator = Vec::with_capacity(n);
+
+                    for part_ord in 0..n {
+                        let part_ord = part_ord as u64;
+                        let part_owner =
+                            self.runtime_adapter.get_part_owner(parent_hash, part_ord)?;
+                        if part_owner == chunk_producer || Some(part_owner) == self.me {
+                            continue;
+                        }
+                        accumulator.push(part_ord);
+                    }
+
+                    accumulator
+                };
+
+                let chosen = candidates
+                    .choose_multiple(
+                        &mut rand::thread_rng(),
+                        cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
+                    )
+                    .cloned()
+                    .collect::<HashSet<_>>();
```

Comment:
It might be better to move into a separate method `get_random_part_ords`

Reply:
Done in https://github.com/nearprotocol/nearcore/pull/2702/commits/86d875e227e8eae48755244bf34d87f337d22497